server/src/main/java/org/elasticsearch/search/vectors/ESDiversifyingChildrenByteKnnVectorQuery.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.vectors;
+
+import org.apache.lucene.index.ByteVectorValues;
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.VectorEncoding;
+import org.apache.lucene.index.VectorSimilarityFunction;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.FieldExistsQuery;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.ScoreDoc;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.search.join.BitSetProducer;
+import org.apache.lucene.search.join.DiversifyingChildrenByteKnnVectorQuery;
+import org.apache.lucene.util.PriorityQueue;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+
+/**
+ * This is an extension of {@link DiversifyingChildrenByteKnnVectorQuery} that gathers the top {@code numChildrenPerParent} children,
+ * not just the single nearest. This is useful for gathering inner hits
+ * It will return the top {@code numChildrenPerParent} children for each parent, this is done as a separate step after
+ * the initial query is executed.
+ */
+public class ESDiversifyingChildrenByteKnnVectorQuery extends DiversifyingChildrenByteKnnVectorQuery {
```

Comment:
You may ask "Ben, this seems really complicated, why don't you extend the InnerHitsContextBuilder thingy in AbstractQueryBuilder for the KnnDocScoreQueryBuilder?"

I will answer, because that would only maybe work for top-level kNN and for it to work (from what I could figure out), `NestedQueryBuilder` would need to know:

 - The original query
 - That one of its children is a kNN query vector (hope that it isn't combined with another bool or something...)
 - Would need special handling vs. the top level kNN

@jimczi maybe you would know better here, but I couldn't see a way to make it work.

Reply:
The key issue is kNN as a query. Mainly because it doesn't get rewritten into `KnnScoreDocQueryBuilder`. We don't gather the global k here, so we just query against the shard, gathering the values there.

Reply:
Maybe we need both. The way I did it currently, and a new `KnnScoreDocQueryBuilder` option that knows the original vector and can rewrite itself appropriately when in a nested context.